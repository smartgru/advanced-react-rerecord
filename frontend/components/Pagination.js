import React from 'react';
import gql from 'graphql-tag';
import { Query } from 'react-apollo';
import { useQuery } from '@apollo/react-hooks';
import Head from 'next/head';
import Link from 'next/link';
import PaginationStyles from './styles/PaginationStyles';
import { perPage } from '../config';
import Error from './ErrorMessage';

const PAGINATION_QUERY = gql`
  query PAGINATION_QUERY {
    _allItemsMeta {
      count
    }
  }
`;

function Pagination(props) {
  const { error, loading, data } = useQuery(PAGINATION_QUERY);
  if (loading) return <p>Loading...</p>;
  if (error) return <Error error={error} />;
  const { count } = data._allItemsMeta;
  const pages = Math.ceil(count / perPage);
  const { page } = props;
  return (
    <PaginationStyles data-test="pagination">
      <Head>
        <title>
          Sick Fits! — Page {page} of {pages}
        </title>
      </Head>
      <Link
        href={{
          pathname: 'items',
          query: { page: page - 1 },
        }}
      >
        <a className="prev" aria-disabled={page <= 1}>
          ← Prev
        </a>
      </Link>
      <p>
        Page {props.page} of
        <span className="totalPages">{pages}</span>!
      </p>
      <p>{count} Items Total</p>
      <Link
        href={{
          pathname: 'items',
          query: { page: page + 1 },
        }}
      >
        <a className="next" aria-disabled={page >= pages}>
          Next →
        </a>
      </Link>
    </PaginationStyles>
  );
}

export default Pagination;
export { PAGINATION_QUERY };
