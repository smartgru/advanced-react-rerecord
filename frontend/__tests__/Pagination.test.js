// import { mount } from 'enzyme';
import { render, fireEvent, screen } from '@testing-library/react';
import wait from 'waait';
// import toJSON from 'enzyme-to-json';
import Router from 'next/router';
import { MockedProvider } from '@apollo/react-testing';
import Pagination, { PAGINATION_QUERY } from '../components/Pagination';

Router.router = {
  push() {},
  prefetch() {},
};

function makeMocksFor(length) {
  return [
    {
      request: { query: PAGINATION_QUERY },
      result: {
        data: {
          _allItemsMeta: {
            count: length,
          },
          itemsConnection: {
            __typename: 'aggregate',
            aggregate: {
              count: length,
              __typename: 'count',
            },
          },
        },
      },
    },
  ];
}

describe('<Pagination/>', () => {
  it('displays a loading message', async () => {
    const { container, queryByText } = render(
      <MockedProvider mocks={makeMocksFor(1)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    expect(queryByText('Loading...')).toBeTruthy();
  });

  it('renders pagination for 18 items', async () => {
    const { container, findByTestId, getByTestId, debug, query } = render(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    // wait for pagination to get past loading
    await findByTestId('pagination');
    expect(getByTestId('totalPages')).toHaveTextContent('5');
    expect(container).toMatchSnapshot();
  });

  it('disables prev button on first page', async () => {
    const { getByText, findByTestId } = render(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={1} />
      </MockedProvider>
    );
    await findByTestId('pagination');
    expect(getByText(/Prev/)).toHaveAttribute('aria-disabled', 'true');
    expect(getByText(/Next/)).toHaveAttribute('aria-disabled', 'false');
  });

  it('disables next button on last page', async () => {
    const { getByText, findByTestId } = render(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={5} />
      </MockedProvider>
    );
    await findByTestId('pagination');
    expect(getByText(/Prev/)).toHaveAttribute('aria-disabled', 'false');
    expect(getByText(/Next/)).toHaveAttribute('aria-disabled', 'true');
  });

  it('enables all buttons on a middle page', async () => {
    const { getByText, findByTestId } = render(
      <MockedProvider mocks={makeMocksFor(18)}>
        <Pagination page={3} />
      </MockedProvider>
    );
    await findByTestId('pagination');
    expect(getByText(/Prev/)).toHaveAttribute('aria-disabled', 'false');
    expect(getByText(/Next/)).toHaveAttribute('aria-disabled', 'false');
  });
});
