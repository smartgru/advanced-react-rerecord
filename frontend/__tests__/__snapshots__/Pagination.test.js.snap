// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Pagination/> renders pagination for 18 items 1`] = `
<div>
  <div
    class="PaginationStyles-aduuar-0 buYoDE"
    data-testid="pagination"
  >
    <a
      aria-disabled="true"
      class="prev"
      href="items?page=0"
    >
      ← Prev
    </a>
    <p>
      Page 
      1
       of
       
      <span
        class="totalPages"
        data-testid="totalPages"
      >
        5
      </span>
    </p>
    <p>
      18
       Items Total
    </p>
    <a
      aria-disabled="false"
      class="next"
      href="items?page=2"
    >
      Next →
    </a>
  </div>
</div>
`;
